### DISCUSSION TITLE
We found pcm-memory.x spawns a lot of threads

### DISCUSSION THREAD

**Discussion by sheperdh** [opcm/pcm#367] [General]
When we launches pcm-memory.x, we found it spawns a lot of threads.
Here is our parameters to launch pcm-memory.x:
`pcm-memory.x -pmm -nc`

From pstack output, we see a lot of threads are just waiting:

> Thread 32 (Thread 0x7efca9929700 (LWP 29789)):
#0  0x00007efcb974ce9d in nanosleep () from /lib64/libpthread.so.0
#1  0x000000000040ff57 in pcm::MySleepMs (delay_ms=<optimized out>) at utils.h:73
#2  pcm::CounterWidthExtender::__lambda23::operator() (__closure=0x10d7c30) at cpucounters.cpp:7528
#3  std::_Bind_simple<pcm::CounterWidthExtender::CounterWidthExtender(pcm::CounterWidthExtender::AbstractRawCounter*, pcm::uint64, pcm::uint32)::__lambda23()>::_M_invoke<> (this=0x10d7c30) at /usr/include/c++/4.8.2/functional:1732
#4  std::_Bind_simple<pcm::CounterWidthExtender::CounterWidthExtender(pcm::CounterWidthExtender::AbstractRawCounter*, pcm::uint64, pcm::uint32)::__lambda23()>::operator() (this=0x10d7c30) at /usr/include/c++/4.8.2/functional:1720
#5  std::thread::_Impl<std::_Bind_simple<pcm::CounterWidthExtender::CounterWidthExtender(pcm::CounterWidthExtender::AbstractRawCounter*, pcm::uint64, pcm::uint32)::__lambda23()> >::_M_run(void) (this=0x10d7c18) at /usr/include/c++/4.8.2/thread:115
#6  0x00007efcb92e3330 in ?? () from /lib64/libstdc++.so.6
#7  0x00007efcb9745ea5 in start_thread () from /lib64/libpthread.so.0
#8  0x00007efcb8a46b0d in clone () from /lib64/libc.so.6
Thread 31 (Thread 0x7efcaa12a700 (LWP 29788)):
#0  0x00007efcb974ce9d in nanosleep () from /lib64/libpthread.so.0
#1  0x000000000040ff57 in pcm::MySleepMs (delay_ms=<optimized out>) at utils.h:73
#2  pcm::CounterWidthExtender::__lambda23::operator() (__closure=0x10d7880) at cpucounters.cpp:7528
#3  std::_Bind_simple<pcm::CounterWidthExtender::CounterWidthExtender(pcm::CounterWidthExtender::AbstractRawCounter*, pcm::uint64, pcm::uint32)::__lambda23()>::_M_invoke<> (this=0x10d7880) at /usr/include/c++/4.8.2/functional:1732
#4  std::_Bind_simple<pcm::CounterWidthExtender::CounterWidthExtender(pcm::CounterWidthExtender::AbstractRawCounter*, pcm::uint64, pcm::uint32)::__lambda23()>::operator() (this=0x10d7880) at /usr/include/c++/4.8.2/functional:1720
#5  std::thread::_Impl<std::_Bind_simple<pcm::CounterWidthExtender::CounterWidthExtender(pcm::CounterWidthExtender::AbstractRawCounter*, pcm::uint64, pcm::uint32)::__lambda23()> >::_M_run(void) (this=0x10d7868) at /usr/include/c++/4.8.2/thread:115
#6  0x00007efcb92e3330 in ?? () from /lib64/libstdc++.so.6
#7  0x00007efcb9745ea5 in start_thread () from /lib64/libpthread.so.0
#8  0x00007efcb8a46b0d in clone () from /lib64/libc.so.6
Thread 30 (Thread 0x7efcaa92b700 (LWP 29787)):
#0  0x00007efcb974ce9d in nanosleep () from /lib64/libpthread.so.0
#1  0x000000000040ff57 in pcm::MySleepMs (delay_ms=<optimized out>) at utils.h:73
#2  pcm::CounterWidthExtender::__lambda23::operator() (__closure=0x10d74d0) at cpucounters.cpp:7528
#3  std::_Bind_simple<pcm::CounterWidthExtender::CounterWidthExtender(pcm::CounterWidthExtender::AbstractRawCounter*, pcm::uint64, pcm::uint32)::__lambda23()>::_M_invoke<> (this=0x10d74d0) at /usr/include/c++/4.8.2/functional:1732
#4  std::_Bind_simple<pcm::CounterWidthExtender::CounterWidthExtender(pcm::CounterWidthExtender::AbstractRawCounter*, pcm::uint64, pcm::uint32)::__lambda23()>::operator() (this=0x10d74d0) at /usr/include/c++/4.8.2/functional:1720
#5  std::thread::_Impl<std::_Bind_simple<pcm::CounterWidthExtender::CounterWidthExtender(pcm::CounterWidthExtender::AbstractRawCounter*, pcm::uint64, pcm::uint32)::__lambda23()> >::_M_run(void) (this=0x10d74b8) at /usr/include/c++/4.8.2/thread:115
#6  0x00007efcb92e3330 in ?? () from /lib64/libstdc++.so.6
#7  0x00007efcb9745ea5 in start_thread () from /lib64/libpthread.so.0
#8  0x00007efcb8a46b0d in clone () from /lib64/libc.so.6

We are wondering what those threads will do, and how we can limit thread number?

**Comment by opcm** [comment]
most of the threads in pcm belong to a pool thread. Threads are used there for parallelization of counter reads and are idling between the sample reads. If you reduce their amount the reading of the counters can become slower.
